aggregator/src/aggregator.rs

```diff
@@ -1742,42 +1655,26 @@ impl VdafOps {
         )
         .get_encoded()
         .map_err(|e| Arc::new(Error::MessageEncode(e)))?;
-        let try_hpke_open = |hpke_keypair: &HpkeKeypair| {
-            hpke::open(
-                hpke_keypair,
-                &HpkeApplicationInfo::new(&Label::InputShare, &Role::Client, task.role()),
-                report.leader_encrypted_input_share(),
-                &input_share_aad,
-            )
-        };
-
-        let global_hpke_keypair =
-            global_hpke_keypairs.keypair(report.leader_encrypted_input_share().config_id());
 
-        let task_hpke_keypair = task
-            .hpke_keys()
-            .get(report.leader_encrypted_input_share().config_id());
+        // Retrieve the HPKE key indicated by the report & verify that it is known.
+        let hpke_keypair =
+            match hpke_keypairs.keypair(report.leader_encrypted_input_share().config_id()) {
+                Some(hpke_keypair) => hpke_keypair,
+                None => {
+                    return Err(reject_report(ReportRejectionReason::OutdatedHpkeConfig(
+                        *report.leader_encrypted_input_share().config_id(),
+                    ))
+                    .await?);
+                }
+            };
 
-        let decryption_result = match (task_hpke_keypair, global_hpke_keypair) {
-            // Verify that the report's HPKE config ID is known.
-            // https://www.ietf.org/archive/id/draft-ietf-ppm-dap-07.html#section-4.4.2-17
-            (None, None) => {
-                return Err(reject_report(ReportRejectionReason::OutdatedHpkeConfig(
-                    *report.leader_encrypted_input_share().config_id(),
-                ))
-                .await?);
-            }
-            (None, Some(global_hpke_keypair)) => try_hpke_open(&global_hpke_keypair),
-            (Some(task_hpke_keypair), None) => try_hpke_open(task_hpke_keypair),
-            (Some(task_hpke_keypair), Some(global_hpke_keypair)) => {
-                try_hpke_open(task_hpke_keypair).or_else(|error| match error {
-                    // Only attempt second trial if _decryption_ fails, and not some
-                    // error in server-side HPKE configuration.
-                    hpke::Error::Hpke(_) => try_hpke_open(&global_hpke_keypair),
-                    error => Err(error),
-                })
-            }
-        };
```

Comment:
Phew, glad this nonsense is gone.

Reply:
Yeah, this was some pretty hairy code, and in general I don't like trial decryption when it can be avoided. I'm glad to see it gone.